langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithStatesTest.java

```diff
@@ -0,0 +1,256 @@
+                    if (customer != null && customer.isValid()) {
+                        System.out.println("Extracted " + customer);
+                        context.setCustomer(customer);
+                    }
+                } catch (Exception ignore) { }
+            } else if (context.getFlight() == null) {
+                try {
+                    Flight flight = flightExtractor.extractData(userMessage);
+                    if (flight != null && flight.isValid()) {
+                        System.out.println("Extracted " + flight);
+                        context.setFlight(flight);
+                    }
+                } catch (Exception ignore) { }
+            }
+
+            String agentMessage = context.isComplete() ?
+                    "Thank you " + context.getCustomer().getFullName() + ", you are eligible for a refund of $" + context.getRefund() :
+                    agent.chat(userMessage);
+            System.out.println("Agent: " + agentMessage);
+        }
+
+        scanner.close();
+    }
+
+    public interface Validated {
+        boolean isValid();
+    }
+
+    public static class Customer implements Validated {
+        public final String firstName;
+        public final String lastName;
+        public final int age;
+
+        public Customer(String firstName, String lastName, int age) {
+            this.firstName = firstName;
+            this.lastName = lastName;
+            this.age = age;
+        }
+
+        @Override
+        public String toString() {
+            return "Customer {" +
+                    " firstName = \"" + firstName + "\"" +
+                    ", lastName = \"" + lastName + "\"" +
+                    ", age = " + age +
+                    " }";
+        }
+
+        public String getFullName() {
+            return firstName + " " + lastName;
+        }
+
+        @Override
+        public boolean isValid() {
+            return firstName != null && !firstName.isEmpty() && age > 0;
+        }
+    }
+
+    public static class Flight implements Validated {
+
+        public final String number;
+        public final int delayInMinutes;
+
+        public Flight(String number, int delayInMinutes) {
+            this.number = number;
+            this.delayInMinutes = delayInMinutes;
+        }
+
+        @Override
+        public String toString() {
+            return "Flight {" +
+                    " number = \"" + number + "\"" +
+                    ", delayInMinutes = " + delayInMinutes +
+                    " }";
+        }
+
+        @Override
+        public boolean isValid() {
+            return number != null && !number.isEmpty() && delayInMinutes > 0;
+        }
+    }
+
+    interface CustomerExtractor {
+
+        @UserMessage("Extract information about a customer from this text '{{it}}'. The response must contain only the JSON with customer's data and without any other sentence.")
```

Comment:
minor: doesn't it work without the "The response must contain only the JSON..." prompt? Similar prompt is already automatically added by AI Services.

BTW, you could enable [JSON mode](https://docs.langchain4j.dev/tutorials/ai-services#json-mode) to make it more reliable.

Reply:
It does, but for some reason the Mistral model that I was using for my local tests works better (or produces the right result more often) when I tell it to reply only with JSON twice ... mysteries of prompt engineering :)